### DISCUSSION TITLE
Factory a mechanism to enforce encapsulation

### DISCUSSION THREAD

**Discussion by OpenTechConsult** [PacktPublishing/Node.js-Design-Patterns-Third-Edition#60] [Q&A]
Hi,

I reached **chapter 7** of the book. I am on the section  _Factory:: A mechanism to enforce encapsulation_.  I can't understand how closure and scope enforce encapsulation.  I can't understand the encapsulation in the given code snippet as an illustration

```js

function createPerson (name) {
  const privateProperties = {}

  const person = {
    setName (name) {
      if (!name) {
        throw new Error('A person must have a name')
      }
      privateProperties.name = name
    },
    getName () {
      return privateProperties.name
    }
  }

  person.setName(name)
  return person
}

const person = createPerson('James Joyce')

console.log(person.getName(), person)
```

**Comment by oscaretu** [comment]
The function createPerson(name) returns an object with 2 methods: setName(name) and getName(). There isn't any other possibility of changing or accessing the value of privateProperties other that using that two methods. Those methods can access the privateProperties because a closure captures the context, and that context includes privateProperties. And the access to the value of privateProperties is encapsulated in the person object.
